assets/src/blocks/Articles/ArticlesFrontend.js

```diff
@@ -0,0 +1,155 @@
+import { Component, Fragment } from '@wordpress/element';
+import { ArticlePreview } from './ArticlePreview';
+
+const { apiFetch } = wp;
+const { addQueryArgs } = wp.url;
+const { __ } = wp.i18n;
+
+export class ArticlesFrontend extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      posts: [],
+      total_pages: 0,
+      page: 0
+    };
+
+    this.loadArticles = this.loadArticles.bind(this);
+  }
+
+  componentDidMount() {
+    this.loadArticles();
+  }
+
+  componentDidUpdate(prevProps) {
+    const { article_count, tags, posts, post_types, ignore_categories } = this.props;
+    if (
+      article_count !== prevProps.article_count ||
+      tags.length !== prevProps.tags.length ||
+      posts.length !== prevProps.posts.length ||
+      post_types.length !== prevProps.post_types.length ||
+      ignore_categories !== prevProps.ignore_categories
+    ) {
+      this.loadArticles();
+    }
+  }
+
+  async loadArticles(page) {
+    const {
+      article_count,
+      posts,
+      tags,
+      post_types,
+      ignore_categories,
+      postType,
+      postId,
+      setTotalPages
+    } = this.props;
+
+    const args = {
+      article_count: article_count && article_count > 0 ? article_count : 0,
+      post_types,
+      posts,
+      tags,
+      ignore_categories
+    };
+
+    if (page) {
+      args.offset = page * article_count;
+    }
+
+    if (postType === 'post') {
+      args.exclude_post_id = postId;
+    }
+
+    const queryArgs = {
+      path: addQueryArgs('/planet4/v1/get-articles', args)
+    };
+
+    const result = await apiFetch(queryArgs);
+
+    let newPosts = [];
+    let total_pages = 0;
+    if (result) {
+      if (page) {
+        newPosts = [...this.state.posts, ...result.recent_posts];
+        total_pages = this.state.total_pages;
+        this.setState({ page });
+      } else {
+        newPosts = result.recent_posts;
+        total_pages = result.total_pages;
+      }
+    }
+    this.setState({
+      posts: newPosts,
+      total_pages
+    });
+    if (setTotalPages) {
+      setTotalPages(total_pages);
+    }
+  }
+
+  render() {
+    const {
+      article_heading,
+      articles_description,
+      read_more_text,
+      read_more_link,
+      button_link_new_tab,
+      isEditing,
+      postType
+    } = this.props;
+
+    const { posts, total_pages, page } = this.state;
+
+    return (
+      <Fragment>
+        <section className="block articles-block">
+          <div className="container">
+            {!isEditing &&
+              <header>
+                <h2 className="page-section-header">{article_heading || __('Latest Articles', 'p4ge')}</h2>
+              </header>
+            }
+            {articles_description && !isEditing &&
+              <div className="page-section-description">{articles_description}</div>
+            }
+            <div className="article-list-section clearfix">
```

Comment:
Looks like this part is the only thing that is rendered when `isEditing` if false. So maybe it's easier to extract this, so that we can include it separately when editing, instead of the entire `ArticlesFrontend`? Then you could remove all the `!isEditing` checks in this component.

Reply:
I'm not sure how to handle this properly, because we load the articles in this component (they're not props) and then the button to load more posts is there too... An idea maybe would be to move the `loadArticles` function to a separate file? Then we could use it both for loading the initial articles (needed both in the frontend and in the editor) and for loading more articles (needed only in the frontend) But then we would need to pass down all the arguments as params for the function instead of using props, I guess that's fine if we use an object 🤔

Reply:
>  An idea maybe would be to move the loadArticles function to a separate file? 

:+1: Then this component will also be a lot simpler.

> Then we could use it both for loading the initial articles (needed both in the frontend and in the editor) and for loading more articles

Why do we need a different call for the initial page and the subsequent ones? I mean, it looks like the count in the first call is only used to see if we need to show the button for the next posts. But we could use also use that for any page.

Reply:
https://reactjs.org/docs/hooks-custom.html#extracting-a-custom-hook